File: 124. Binary Tree Maximum Path Sum-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        res = -inf

        def dfs(root):
            if root is None:
                return 0
            x = root.val
            l_len = dfs(root.left)
            r_len = dfs(root.right)
            t_len = x
            if l_len > 0:
                t_len += l_len
            if r_len > 0:
                t_len += r_len
            nonlocal res
            res = max(res, t_len)
            return max(x, max(x + l_len, x + r_len))

        dfs(root)
        return res

File: 124. Binary Tree Maximum Path Sum-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
  int res = INT_MIN / 2;
  int maxPathSum(TreeNode *root) {
    dfs(root);

    return res;
  }

  int dfs(TreeNode *root) {
    if (!root)
      return 0;
    int left = dfs(root->left);
    int right = dfs(root->right);
    int x = root->val;
    int cur = x;
    cur += left > 0 ? left : 0;
    cur += right > 0 ? right : 0;
    res = max(res, cur);
    return max(x, max(x + left, x + right));
  }
};

class Solution {
public:
  int res = INT_MIN;
  int maxPathSum(TreeNode *root) {
    helper(root);
    return res;
  }

  int helper(TreeNode *root) {
    if (!root)
      return 0;
    int mx_l = helper(root->left);
    int mx_r = helper(root->right);
    res = max(res, root->val + mx_l + mx_r);
    return max(0, root->val + max(mx_l, mx_r));
  }
};